"YouTube is in Fullscreen” banner is not liquid glass

Phinny01 opened the pull request:
#### bc0503a149a3f710a8322a3b6c860a848c3bad6a
<pre>
&quot;YouTube is in Fullscreen” banner is not liquid glass
<a href="https://bugs.webkit.org/show_bug.cgi?id=298749">https://bugs.webkit.org/show_bug.cgi?id=298749</a>
<a href="https://rdar.apple.com/149900187">rdar://149900187</a>

Reviewed by Abrar Rahman Protyasha.

Applies UIGlassEffect to the fullscreen dismissal banner(&quot;YouTube is in fullscreen&quot;) to achieve
the liquid glass style for the new OS, while maintaining a fallback for the old OS.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController loadView]):
(-[WKFullScreenViewController viewDidLayoutSubviews]):

Canonical link: <a href="https://commits.webkit.org/300014@main">https://commits.webkit.org/300014@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/216589ecd1316da5f7ab1325bd6107758caff682

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121025 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40721 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31379 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127443 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73105 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41423 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49300 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91919 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61149 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123977 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33068 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108480 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72607 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32092 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71032 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102573 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26762 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130297 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47952 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36432 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100529 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48320 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104647 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100431 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25464 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45836 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23881 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/44640 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47810 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53523 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47281 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50628 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48965 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -829,15 +829,21 @@ - (void)loadView
-    _banner = WTFMove(banner);
+#if HAVE(UI_GLASS_EFFECT)
+    RetainPtr<UIVisualEffect> glassEffect = adoptNS([[UIGlassEffect alloc] init]);
+    RetainPtr<UIVisualEffectView> bannerEffectView = adoptNS([[UIVisualEffectView alloc] initWithEffect:glassEffect.get()]);
```
You can drop the template arg because it can be inferred from the resulting pointer.
```suggestion
    RetainPtr bannerEffectView = adoptNS([[UIVisualEffectView alloc] initWithEffect:glassEffect.get()]);
```

Phinny01 replied:
right, I have removed it.

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -829,15 +829,21 @@ - (void)loadView
+
+    [bannerEffectView.get().contentView addSubview:_bannerLabel.get()];
+
+    bannerEffectView.get().clipsToBounds = YES;
```
Nit:


```suggestion
    [[bannerEffectView contentView] addSubview:_bannerLabel.get()];

    [bannerEffectView setClipsToBounds:YES];
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -846,6 +852,14 @@ - (void)loadView
+        [_bannerLabel.get().leadingAnchor constraintEqualToAnchor:_banner.get().contentView.leadingAnchor constant:16],
+        [_bannerLabel.get().trailingAnchor constraintEqualToAnchor:_banner.get().contentView.trailingAnchor constant:-16],
+        [_bannerLabel.get().topAnchor constraintEqualToAnchor:_banner.get().contentView.topAnchor constant:16],
+        [_bannerLabel.get().bottomAnchor constraintEqualToAnchor:_banner.get().contentView.bottomAnchor constant:-16]
```
nit: I suppose these reflect the edge insets on `_bannerLabel`, maybe use a common constant instead?

Phinny01 replied:
Yes that makes it better.

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -941,6 +984,11 @@ - (void)viewDidLayoutSubviews
+
+#if HAVE(UI_GLASS_EFFECT)
+    if (_banner)
+        _banner.get().layer.cornerRadius = _banner.get().layer.bounds.size.height / 2;
```
One more style suggestion: prefer `[foo bar]` over `foo.get().bar`
```suggestion
        [[_banner layer] setCornerRadius:[[_banner layer].bounds.size.height / 2]];
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -827,14 +831,73 @@ - (void)loadView
+        _bannerTapToDismissRecognizer = nil;
+    }
+
+    static const CGFloat bannerLabelInset = 16;
```
This can be `constexpr`
```suggestion
    static constexpr CGFloat bannerLabelInset = 16;
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -827,14 +831,73 @@ - (void)loadView
+#if HAVE(UI_GLASS_EFFECT)
+    _bannerLabel = adoptNS([[_WKInsetLabel alloc] initWithFrame:CGRectZero]);
+#else
     _bannerLabel = adoptNS([[_WKInsetLabel alloc] initWithFrame:CGRectMake(0, 0, 100, 100)]);
```
nit: You can have one call to `[_WKInsetLabel alloc]` if you structure this way:


```cpp
#if HAVE(UI_GLASS_EFFECT)
    auto labelFrame = CGRectZero;
#else
    auto labelFrame = CGRectMake(0, 0, 100, 100);
#endif
    _bannerLabel = adoptNS([[_WKInsetLabel alloc] initWithFrame:labelFrame]);
```

aprotyas commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -846,6 +909,7 @@ - (void)loadView
     [_banner setTranslatesAutoresizingMaskIntoConstraints:NO];
 
     [_animatingView addSubview:_banner.get()];
+#endif
```
All of this seems to be common code between the glass effect vs legacy paths. Is there any way you could avoid the duplication?

```objc
    _banner = WTFMove(banner);

    _bannerTapToDismissRecognizer = adoptNS([[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(_bannerDismissalRecognized:)]);
    [_bannerTapToDismissRecognizer setDelegate:self];
    [_banner addGestureRecognizer:_bannerTapToDismissRecognizer.get()];

    [_banner setTranslatesAutoresizingMaskIntoConstraints:NO];

    [_animatingView addSubview:_banner.get()];
```

aprotyas reviewed (commented):
This is looking good, thanks for iterating!

Commit:
"YouTube is in Fullscreen” banner is not liquid glass
https://bugs.webkit.org/show_bug.cgi?id=298749
rdar://149900187

Reviewed by Abrar Rahman Protyasha.

Applies UIGlassEffect to the fullscreen dismissal banner("YouTube is in fullscreen") to achieve
the liquid glass style for the new OS, while maintaining a fallback for the old OS.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController loadView]):
(-[WKFullScreenViewController viewDidLayoutSubviews]):

Canonical link: https://commits.webkit.org/300014@main
